batchskl: panic due to index out of range

Reported in cockroachdb/cockroach#69906.

We're hitting the following in `batchskl.Skiplist`:

```console
panic.go:679: runtime error: index out of range [4294967264] with length 0
--
runtime.boundsError
panic.go:679: *withstack.withStack (top exception)

/usr/local/go/src/runtime/panic.go in runtime.gopanic at line 679
/usr/local/go/src/runtime/panic.go in runtime.goPanicIndex at line 75
vendor/github.com/cockroachdb/pebble/internal/batchskl/skl.go in github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).node at line 296
vendor/github.com/cockroachdb/pebble/internal/batchskl/skl.go in github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).newNode at line 265
vendor/github.com/cockroachdb/pebble/internal/batchskl/skl.go in github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).Add at line 233
vendor/github.com/cockroachdb/pebble/batch.go in github.com/cockroachdb/pebble.(*Batch).Delete at line 528
pkg/storage/pebble_batch.go in pkg/storage.(*pebbleBatch).Clear at line 240
pkg/kv/kvserver/replica_raft.go in pkg/kv/kvserver.handleTruncatedStateBelowRaft at line 1696
pkg/kv/kvserver/replica_application_state_machine.go in pkg/kv/kvserver.(*replicaAppBatch).runPreApplyTriggersAfterStagingWriteBatch at line 702
pkg/kv/kvserver/replica_application_state_machine.go in pkg/kv/kvserver.(*replicaAppBatch).Stage at line 484
pkg/kv/kvserver/apply/cmd.go in pkg/kv/kvserver/apply.mapCmdIter at line 175
pkg/kv/kvserver/apply/task.go in pkg/kv/kvserver/apply.(*Task).applyOneBatch at line 280
pkg/kv/kvserver/apply/task.go in pkg/kv/kvserver/apply.(*Task).ApplyCommittedEntries at line 247
pkg/kv/kvserver/replica_raft.go in pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked at line 765
pkg/kv/kvserver/replica_raft.go in pkg/kv/kvserver.(*Replica).handleRaftReady at line 425
pkg/kv/kvserver/store_raft.go in pkg/kv/kvserver.(*Store).processReady at line 484
pkg/kv/kvserver/scheduler.go in pkg/kv/kvserver.(*raftScheduler).worker at line 226
pkg/kv/kvserver/scheduler.go in pkg/kv/kvserver.(*raftScheduler).Start.func2 at line 166
pkg/util/stop/stopper.go in pkg/util/stop.(*Stopper).RunWorker.func1 at line 222
/usr/local/go/src/runtime/asm_amd64.s in runtime.goexit at line 1357
```

This corresponds to Pebble SHA [`c3ef93f9a9edfa7a8ad8d897df6719a7deeba71d`](https://github.com/cockroachdb/pebble/blob/c3ef93f9a9edfa7a8ad8d897df6719a7deeba71d).

The relevant code is around here: https://github.com/cockroachdb/pebble/blob/c3ef93f9a9edfa7a8ad8d897df6719a7deeba71d/internal/batchskl/skl.go#L274-L297

Based on the stacktrace, I think what might be happening is the following:

- large skiplist with `len(nodes)=4294967264` (value of `offset`)
- we compute `newSize` as `4294967264 + size`
- we then resize `s.nodes` to `s.nodes[:newSize]`
- the stacktrace tells us the length of nodes was 0
- working backwards, size was most likely 32, given that `(4294967264 + 32) mod 2^32 == 0`

Reproducer:

```go
func Test(t *testing.T) {
	opts := &Options{
		FS: vfs.NewMem(),
	}
	db, err := Open("", opts)
	if err != nil {
		panic(err)
	}
	defer db.Close()

	b := db.NewIndexedBatch()
	for i := 0; i < (1 << 32); i++ {
		s := strconv.Itoa(i)
		if err = b.Set([]byte(s), []byte{}, nil); err != nil {
			panic(err)
		}
		if (i % (1 << 25)) == 0 {
			fmt.Printf("added %d items to batch\n", i)
		}
	}

	if err = b.Commit(nil); err != nil {
		panic(err)
	}
	defer b.Close()
}
```

```console
--- FAIL: Test (34.52s)
panic: runtime error: index out of range [4294967272] with length 8 [recovered]
	panic: runtime error: index out of range [4294967272] with length 8

goroutine 19 [running]:
testing.tRunner.func1.2(0x46ad0e0, 0xc0001d0228)
	/nix/store/l7w2b12gyhznbv6df50n4j9y3x8ga659-go-1.16.7/share/go/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc000186a80)
	/nix/store/l7w2b12gyhznbv6df50n4j9y3x8ga659-go-1.16.7/share/go/src/testing/testing.go:1146 +0x4b6
panic(0x46ad0e0, 0xc0001d0228)
	/nix/store/l7w2b12gyhznbv6df50n4j9y3x8ga659-go-1.16.7/share/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).node(...)
	/Users/nickt/Development/pebble/internal/batchskl/skl.go:296
github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).newNode(0xc0001d47f0, 0x4d2ec68200000001, 0x4d2ec68d4d2ec684, 0x3131373136383531, 0x3131373136383531)
	/Users/nickt/Development/pebble/internal/batchskl/skl.go:265 +0x170
github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).Add(0xc0001d47f0, 0xc04d2ec682, 0x0, 0xc00028bf01)
	/Users/nickt/Development/pebble/internal/batchskl/skl.go:233 +0x29b
github.com/cockroachdb/pebble.(*Batch).Set(0xc0001d46e0, 0xc00028bf00, 0x9, 0x20, 0xc00028bef7, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/nickt/Development/pebble/batch.go:514 +0xde
github.com/cockroachdb/pebble.Test(0xc000186a80)
	/Users/nickt/Development/pebble/test_test.go:26 +0x23b
testing.tRunner(0xc000186a80, 0x470c380)
	/nix/store/l7w2b12gyhznbv6df50n4j9y3x8ga659-go-1.16.7/share/go/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/nix/store/l7w2b12gyhznbv6df50n4j9y3x8ga659-go-1.16.7/share/go/src/testing/testing.go:1238 +0x2b3
```

--- a/batch.go
+++ b/batch.go
@@ -64,13 +64,13 @@ type DeferredBatchOp struct {
 // have been filled into Key and Value. Not calling Finish or not
 // copying/encoding keys will result in an incomplete index, and calling Finish
 // twice may result in a panic.
-func (d DeferredBatchOp) Finish() {
+func (d DeferredBatchOp) Finish() error {
 	if d.index != nil {
 		if err := d.index.Add(d.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
+	return nil
 }
 
 // A Batch is a sequence of Sets, Merges, Deletes, DeleteRanges, RangeKeySets,
@@ -394,8 +394,7 @@ func (b *Batch) Apply(batch *Batch, _ *WriteOptions) error {
 					err = b.index.Add(uint32(offset))
 				}
 				if err != nil {
-					// We never add duplicate entries, so an error should never occur.
-					panic(err)
+					return err
 				}
 			}
 			b.memTableSize += memTableEntrySize(len(key), len(value))
@@ -512,8 +511,7 @@ func (b *Batch) Set(key, value []byte, _ *WriteOptions) error {
 	// in go1.13 will remove the need for this.
 	if b.index != nil {
 		if err := b.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -542,8 +540,7 @@ func (b *Batch) Merge(key, value []byte, _ *WriteOptions) error {
 	// in go1.13 will remove the need for this.
 	if b.index != nil {
 		if err := b.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -569,8 +566,7 @@ func (b *Batch) Delete(key []byte, _ *WriteOptions) error {
 	// in go1.13 will remove the need for this.
 	if b.index != nil {
 		if err := b.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -597,8 +593,7 @@ func (b *Batch) SingleDelete(key []byte, _ *WriteOptions) error {
 	// in go1.13 will remove the need for this.
 	if b.index != nil {
 		if err := b.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -628,8 +623,7 @@ func (b *Batch) DeleteRange(start, end []byte, _ *WriteOptions) error {
 	// in go1.13 will remove the need for this.
 	if deferredOp.index != nil {
 		if err := deferredOp.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -679,8 +673,7 @@ func (b experimentalBatch) RangeKeySet(start, end, suffix, value []byte, _ *Writ
 	// Manually inline DeferredBatchOp.Finish().
 	if deferredOp.index != nil {
 		if err := deferredOp.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -719,8 +712,7 @@ func (b experimentalBatch) RangeKeyUnset(start, end, suffix []byte, _ *WriteOpti
 	// Manually inline DeferredBatchOp.Finish()
 	if deferredOp.index != nil {
 		if err := deferredOp.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil
@@ -740,8 +732,7 @@ func (b experimentalBatch) RangeKeyDelete(start, end []byte, _ *WriteOptions) er
 	// Manually inline DeferredBatchOp.Finish().
 	if deferredOp.index != nil {
 		if err := deferredOp.index.Add(deferredOp.offset); err != nil {
-			// We never add duplicate entries, so an error should never occur.
-			panic(err)
+			return err
 		}
 	}
 	return nil

--- a/internal/batchskl/skl.go
+++ b/internal/batchskl/skl.go
@@ -67,15 +67,23 @@ import (
 )
 
 const (
-	maxHeight   = 20
-	maxNodeSize = int(unsafe.Sizeof(node{}))
-	linksSize   = int(unsafe.Sizeof(links{}))
+	maxHeight    = 20
+	maxNodeSize  = int(unsafe.Sizeof(node{}))
+	linksSize    = int(unsafe.Sizeof(links{}))
+	maxNodesSize = math.MaxUint32
 )
 
-// ErrExists indicates that a duplicate record was inserted. This should never
-// happen for normal usage of batchskl as every key should have a unique
-// sequence number.
-var ErrExists = errors.New("record with this key already exists")
+var (
+	// ErrExists indicates that a duplicate record was inserted. This should never
+	// happen for normal usage of batchskl as every key should have a unique
+	// sequence number.
+	ErrExists = errors.New("record with this key already exists")
+
+	// ErrTooManyRecords is a sentinel error returned when the size of the raw
+	// nodes slice exceeds the maximum allowed size (currently 1 << 32 - 1). This
+	// corresponds to ~117 M skiplist entries.
+	ErrTooManyRecords = errors.New("too many records")
+)
 
 type links struct {
 	next uint32
@@ -171,9 +179,17 @@ func (s *Skiplist) Init(storage *[]byte, cmp base.Compare, abbreviatedKey base.A
 		s.nodes = make([]byte, 0, initBufSize)
 	}
 
-	// Allocate head and tail nodes.
-	s.head = s.newNode(maxHeight, 0, 0, 0, 0)
-	s.tail = s.newNode(maxHeight, 0, 0, 0, 0)
+	// Allocate head and tail nodes. While allocating a new node can fail, in the
+	// context of initializing the skiplist we consider it unrecoverable.
+	var err error
+	s.head, err = s.newNode(maxHeight, 0, 0, 0, 0)
+	if err != nil {
+		panic(err)
+	}
+	s.tail, err = s.newNode(maxHeight, 0, 0, 0, 0)
+	if err != nil {
+		panic(err)
+	}
 
 	// Link all head/tail levels together.
 	headNode := s.node(s.head)
@@ -230,7 +246,10 @@ func (s *Skiplist) Add(keyOffset uint32) error {
 	// We always insert from the base level and up. After you add a node in base
 	// level, we cannot create a node in the level above because it would have
 	// discovered the node in the base level.
-	nd := s.newNode(height, keyOffset, keyStart, keyEnd, abbreviatedKey)
+	nd, err := s.newNode(height, keyOffset, keyStart, keyEnd, abbreviatedKey)
+	if err != nil {
+		return err
+	}
 	newNode := s.node(nd)
 	for level := uint32(0); level < height; level++ {
 		next := spl[level].next
@@ -255,23 +274,26 @@ func (s *Skiplist) NewIter(lower, upper []byte) Iterator {
 }
 
 func (s *Skiplist) newNode(height,
-	offset, keyStart, keyEnd uint32, abbreviatedKey uint64) uint32 {
+	offset, keyStart, keyEnd uint32, abbreviatedKey uint64) (uint32, error) {
 	if height < 1 || height > maxHeight {
 		panic("height cannot be less than one or greater than the max height")
 	}
 
 	unusedSize := (maxHeight - int(height)) * linksSize
-	nodeOffset := s.alloc(uint32(maxNodeSize - unusedSize))
+	nodeOffset, err := s.alloc(uint32(maxNodeSize - unusedSize))
+	if err != nil {
+		return 0, err
+	}
 	nd := s.node(nodeOffset)
 
 	nd.offset = offset
 	nd.keyStart = keyStart
 	nd.keyEnd = keyEnd
 	nd.abbreviatedKey = abbreviatedKey
-	return nodeOffset
+	return nodeOffset, nil
 }
 
-func (s *Skiplist) alloc(size uint32) uint32 {
+func (s *Skiplist) alloc(size uint32) (uint32, error) {
 	offset := len(s.nodes)
 
 	// We only have a need for memory up to offset + size, but we never want
@@ -282,6 +304,19 @@ func (s *Skiplist) alloc(size uint32) uint32 {
 		if allocSize < minAllocSize {
 			allocSize = minAllocSize
 		}
+		// Cap the allocation at the max allowed size to avoid wasted capacity.
+		if allocSize > maxNodesSize {
+			// The new record may still not fit within the allocation, in which case
+			// we return early with an error. This avoids the panic below when we
+			// resize the slice. It also avoids the allocation and copy.
+			if uint64(offset)+uint64(size) > maxNodesSize {
+				return 0, errors.Wrapf(ErrTooManyRecords,
+					"alloc of new record (size=%d) would overflow uint32 (current size=%d)",
+					uint64(offset)+uint64(size), offset,
+				)
+			}
+			allocSize = maxNodesSize
+		}
 		tmp := make([]byte, len(s.nodes), allocSize)
 		copy(tmp, s.nodes)
 		s.nodes = tmp
@@ -289,7 +324,7 @@ func (s *Skiplist) alloc(size uint32) uint32 {
 
 	newSize := uint32(offset) + size
 	s.nodes = s.nodes[:newSize]
-	return uint32(offset)
+	return uint32(offset), nil
 }
 
 func (s *Skiplist) node(offset uint32) *node {

--- a/internal/batchskl/skl_test.go
+++ b/internal/batchskl/skl_test.go
@@ -24,6 +24,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble/internal/base"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/exp/rand"
@@ -213,6 +214,24 @@ func TestSkiplistAdd(t *testing.T) {
 	require.Nil(t, l.Add(d.add("00002")))
 	require.Equal(t, 6, length(l))
 	require.Equal(t, 6, lengthRev(l))
+}
+
+func TestSkiplistAdd_Overflow(t *testing.T) {
+	// Regression test for cockroachdb/pebble#1258. The length of the nodes buffer
+	// cannot exceed the maximum allowable size.
+	d := &testStorage{}
+	l := newTestSkiplist(d)
+
+	// Simulate a full nodes slice. This speeds up the test significantly, as
+	// opposed to adding data to the list.
+	l.nodes = make([]byte, maxNodesSize)
+
+	// Adding a new node to the list would overflow the nodes slice. Note that it
+	// is the size of a new node struct that is relevant here, rather than the
+	// size of the data being added to the list.
+	err := l.Add(d.add("too much!"))
+	require.Error(t, err)
+	require.True(t, errors.Is(err, ErrTooManyRecords))
 }
 
 // TestIteratorNext tests a basic iteration over all nodes from the beginning.
@@ -480,7 +499,8 @@ func BenchmarkIterNext(b *testing.B) {
 	for len(d.data)+20 < cap(d.data) {
 		key := randomKey(rng, buf[:])
 		offset := d.addBytes(key)
-		_ = l.Add(offset)
+		err := l.Add(offset)
+		require.NoError(b, err)
 	}
 
 	it := l.NewIter(nil, nil)
@@ -504,7 +524,8 @@ func BenchmarkIterPrev(b *testing.B) {
 	for len(d.data)+20 < cap(d.data) {
 		key := randomKey(rng, buf[:])
 		offset := d.addBytes(key)
-		_ = l.Add(offset)
+		err := l.Add(offset)
+		require.NoError(b, err)
 	}
 
 	it := l.NewIter(nil, nil)

